### PR TITLE
feat(compiler): check serializability of json elements

### LINKED ISSUE #1683: Validate Json Types are legal
When creating a Json in Wing the types allowed should be limited. IE right now we can do something like this:
```js
let b = new cloud.Bucket();
let arr = Json [b,b,b];
```

however we should get an error compiling this like:
```
let arr = Json [b,b,b];
//              ^ type cloud.Bucket is not serializable 
Only serializable types should be allowed in Json objects. Maybe we can solve this using interface system.

### PR DESCRIPTION
**Pull request by YarivLevy81**
Should close #1683.

Added validation for serializability of Json elements.

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

**Comment by YarivLevy81** [comment]
Missing some more tests, will add them

File: examples/tests/invalid/json.w [hunk 1]
```diff
@@ -20,7 +20,10 @@ let forever_json = Json {a: "hello"};
...
-// let bkt = new cloud.Bucket();
-// let j_arr = Json [bkt];
-//                   ^^^ type cloud.bucket is not serializable 
\ No newline at end of file
+let bkt = new cloud.Bucket();
+let j_arr = Json [bkt];
+//                ^^^ Expected "Json" elements to be serializable, but got "Bucket" which is not serializable
+
+let json_obj = Json { boom: bkt };
+//                ^^^ Expected "Json" elements to be serializable, but got "Bucket" which is not serializable
```
Comment (hasanaburayyan) [inline examples/tests/invalid/json.w:28]: The tics should be under `bkt`

File: libs/wingc/src/type_check.rs [hunk 2]
```diff
@@ -636,6 +636,28 @@ impl TypeRef {
 			_ => false,
 		}
 	}
+
+	pub fn is_serializable(&self) -> bool {
+		match **self {
+			Type::Resource(_) => false,
+			Type::Enum(_) => true,
```
Comment (hasanaburayyan) [inline libs/wingc/src/type_check.rs:643]: For now lets leave Enum as false
```suggestion
			Type::Enum(_) => false,
```

File: libs/wingc/src/type_check.rs [hunk 3]
```diff
@@ -636,6 +636,28 @@ impl TypeRef {
...
+			Type::Json => true,
+			Type::Array(v) => v.is_serializable(),
+			Type::Map(v) => v.is_serializable(),
+			Type::Set(v) => v.is_serializable(),
+			Type::Struct(_) => self.as_struct().unwrap().env.iter(true).all(|(_, v, _)| {
+				v.as_variable()
+					.expect("Expected struct field to be a variable in the struct env")
+					.type_
+					.is_serializable()
+			}),
```
Comment (hasanaburayyan) [inline libs/wingc/src/type_check.rs:657]: As well lets keep structs false for now
```suggestion
			Type::Struct(_) => false,
```

File: libs/wingc/src/type_check.rs [hunk 4]
```diff
@@ -1376,6 +1392,31 @@ impl<'a> TypeChecker<'a> {
...
+				self.expr_error(
+					exp,
+					format!(
+						"Expected \"Json\" elements to be serializable, but got \"{}\" which is not serializable",
+						actual_type
+					),
+				);
+			}
+
+			return actual_type;
```
Comment (hasanaburayyan) [inline libs/wingc/src/type_check.rs:1414]: We can get rid of a level of nesting with this re-org. Score us some never-nesting points with @Chriscbr :grin:
```suggestion
		if !context.in_json {
			return self.validate_type(actual_type, expected_type, exp);
		}

		if !actual_type.is_serializable() {
			self.expr_error(
				exp,
				format!(
					"Expected \"Json\" elements to be serializable, but got \"{}\" which is not serializable",
					actual_type
				),
			);
		}
		actual_type
```

File: apps/jsii-docgen/test/docgen/view/__snapshots__/markdown.test.ts.snap [hunk 5]
```diff
@@ -21,49 +21,49 @@ new GreeterBucket(props?: BucketProps)
 
 | **Name** | **Type** | **Description** |
 | --- | --- | --- |
-| <code><a href=\\"#custom-construct-library.GreeterBucket.Initializer.parameter.props\\">props</a></code> | <code><a href=\\"#custom-@aws-cdk/aws-s3.BucketProps\\">BucketProps</a></code> | *No description.* |
+| <code><a href=\\"/packages/construct-library/v/0.0.0/api/GreeterBucket.Initializer.parameter.props?lang=python\\">props</a></code> | <code><a href=\\"/packages/@aws-cdk/aws-s3/v/99.99.99/api/BucketProps?lang=python\\">BucketProps</a></code> | *No description.* |
```
Comment (hasanaburayyan) [inline apps/jsii-docgen/test/docgen/view/__snapshots__/markdown.test.ts.snap:24]: hmm, these changes dont seem intentional. Im not sure why there is all this version 0.0.0 and `lang=python` updates. @Chriscbr or @MarkMcCulloh any ideas?

**Review by hasanaburayyan — CHANGES_REQUESTED** [review]
💪 🚀 Awesome stuff, couple minor change requests. Let me know what you think.

File: libs/wingc/src/type_check.rs [hunk 6]
```diff
@@ -636,6 +636,28 @@ impl TypeRef {
...
 		}
 	}
+
+	pub fn is_serializable(&self) -> bool {
+		match **self {
+			Type::Resource(_) => false,
+			Type::Enum(_) => true,
+			Type::Number => true,
+			Type::String => true,
+			Type::Duration => true,
```
Comment (ekeren) [inline libs/wingc/src/type_check.rs:646]: duration should not be supported 
```suggestion
			Type::Duration => false,
```

File: examples/tests/invalid/json.w (hunk 1 above)
Comment (ekeren) [reply examples/tests/invalid/json.w:28]: We do not support any serializable value in Json we only support str,num, bool and arrays with these values (or another json)  - https://docs.winglang.io/reference/spec#114-json-type

I would change the error message and remove serializable 

//                ^^^ Expected "Json" elements to be Json Value (https://www.json.org/json-en.html)

File: libs/wingc/src/type_check.rs [hunk 7]
```diff
@@ -636,6 +636,28 @@ impl TypeRef {
...
+			Type::Boolean => true,
+			Type::Json => true,
+			Type::Array(v) => v.is_serializable(),
+			Type::Map(v) => v.is_serializable(),
```
Comment (ekeren) [inline libs/wingc/src/type_check.rs:650]: I think that Map should not be valid. 

and maybe `is_serializable` is the wrong term to use. Enum is serializable but is not a json_value
maybe something like `is_legal_json_value`

File: apps/jsii-docgen/test/docgen/view/__snapshots__/markdown.test.ts.snap (hunk 5 above)
Comment (Chriscbr) [reply apps/jsii-docgen/test/docgen/view/__snapshots__/markdown.test.ts.snap:24]: Not sure off the top of my head - let's try to just remove it from the branch for now

File: libs/wingc/src/type_check.rs (hunk 7 above)
Comment (hasanaburayyan) [reply libs/wingc/src/type_check.rs]: We need to leave map as valid. As for the name change I think thats fine.

**Review by hasanaburayyan — COMMENTED** [review]
@YarivLevy81 thanks for getting those changes in! 🎉 

Lets uncommit the changes to `apps/jsii-docgen/test/docgen/view/__snapshots__/markdown.test.ts.snap` and I had 1 other suggestion. 

Happy to help if you need any

File: libs/wingc/src/type_check.rs [hunk 8]
```diff
@@ -637,6 +637,23 @@ impl TypeRef {
...
+			Type::Boolean => true,
+			Type::Json => true,
+			Type::Array(v) => v.is_json_legal_value(),
+			Type::Map(v) => v.is_json_legal_value(),
+			Type::Set(v) => v.is_json_legal_value(),
+			Type::Struct(_) => false,
+			_ => false,
+		}
+	}
 }
```
[the remark this answers is not in the document]
Comment (YarivLevy81) [reply libs/wingc/src/type_check.rs]: I think we're good now 😃

[the remark this answers is not in the document]
Comment (staycoolcall911) [reply libs/wingc/src/type_check.rs]: Love this change!

[the remark this answers is not in the document]
Comment (YarivLevy81) [reply libs/wingc/src/type_check.rs]: Merge? @hasanaburayyan

**Comment by YarivLevy81** [comment]
> Thank you for contributing! Your pull request will be updated from main and then merged automatically (do not update manually, and be sure to [allow changes to be pushed to your fork](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)).

I think the bot won't embark the change @hasanaburayyan 😒

**Pull request merged**